File: 113. Path Sum II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        if not root:
            return []
        results = []
        self.helper(root, sum, [], results)
        return results
        
    def helper(self, root, sum, result, results):
        if not root.left and not root.right:
            if sum == root.val:
                results.append(result + [root.val])
            return
            
        sum -= root.val
        if root.left:
            self.helper(root.left, sum, result+[root.val], results)
        if root.right:
            self.helper(root.right, sum, result+[root.val], results)
            
    # BFS + queue    
    def pathSum3(self, root, sum): 
        if not root:
            return []
        res = []
        queue = [(root, root.val, [root.val])]
        while queue:
            curr, val, ls = queue.pop(0)
            if not curr.left and not curr.right and val == sum:
                res.append(ls)
            if curr.left:
                queue.append((curr.left, val+curr.left.val, ls+[curr.left.val]))
            if curr.right:
                queue.append((curr.right, val+curr.right.val, ls+[curr.right.val]))
        return res
    
    # DFS + stack I  
    def pathSum4(self, root, sum): 
        if not root:
            return []
        res = []
        stack = [(root, sum-root.val, [root.val])]
        while stack:
            curr, val, ls = stack.pop()
            if not curr.left and not curr.right and val == 0:
                res.append(ls)
            if curr.right:
                stack.append((curr.right, val-curr.right.val, ls+[curr.right.val]))
            if curr.left:
                stack.append((curr.left, val-curr.left.val, ls+[curr.left.val]))
        return res 

File: 113. Path Sum II.java
import java.util.ArrayList;
import java.util.List;

import javax.swing.tree.TreeNode;

/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {
    /**
     * backtracking
     * Points to notice:
     * 1. for java, we need to remove the point we add after the recursion
     * 2. when we add new list into the result, we need to make a deep copy (new ArrayList<>(temp))
     * @param root
     * @param sum
     * @return
     */
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        if(root == null){
            return new ArrayList<>();
        }
        List<List<Integer>> res = new ArrayList<>();
        helper(root,sum, new ArrayList<>(), res);
        return res;
    }
    
    public void helper(TreeNode root, int sum, List<Integer> temp, List<List<Integer>> res){
        if(root == null){
            return;
        }
        temp.add(root.val);
        
        if(root.left == null && root.right == null && root.val == sum){
            res.add(new ArrayList<>(temp));
            temp.remove(temp.size() -1);
            return;
        }
        helper(root.left, sum - root.val, temp, res);
        helper(root.right, sum - root.val, temp, res);
        temp.remove(temp.size() -1);
    }
}